core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2890,6 +2893,9 @@ public void run() {
         LogFactory.releaseAll();
 
         theInstance = null;
+        if (JenkinsJVM.isJenkinsJVM()) {
+            JenkinsJVMAccess._setJenkinsJVM(oldJenkinsJVM);
```

Comment:
BTW is there really any use case for calling `_setJenkinsJVM(false)`? I mean, if the intended purpose of `JenkinsJVM` is to differentiate master from slave VMs, once a master always a master, regardless of whether it is shutting down, restarting, whatever. You potentially need a null check on `getInstance()` anyway, if your code might be run at weird times.